### PR TITLE
Rum 22010 write new collectors

### PR DESCRIPTION
**Pull request by robertdumitrescu**

### CODE REVIEW

File: examples/Vanilla/index.html [hunk 1]
```diff
@@ -13,6 +13,8 @@ <h1>Examples</h1>
     <li><a href="OnLoad/index.html">On Load Page</a></li>
     <li><a href="Error/index.html">Error Page</a></li>
     <li><a href="CustomCollector/index.html">Custom Collector Page</a></li>
+    <li><a href="WebBasedCollectorsCollection/index.html">Web Based Collectors Collection Page</a></li>
+    <li><a href="WebBasedCollectorsCollection/index.html">Web Based Collectors Page</a></li>
```
Comment (yghern) [inline examples/Vanilla/index.html]: why these are identical? shouldn't it be WebBasedCollector and WebBasedCollectorCollection?

Comment (robertdumitrescu) [reply examples/Vanilla/index.html]: One of them is for when you use them manually one by one. The other one is for when you want to use a pre made collection with all the web focused ones.

**Review by yghern — APPROVED** [review]
Looking good

File: README.md [hunk 2]
```diff
@@ -114,10 +114,14 @@ A producer is a `class` which executes all the `Collectors` provided. It expects
...
 A `function`, `object` or `class` with an `async` prepare method. The prepare method is given an `event` and is expected to return an augmented `event`.
 
-#### defaultCollectors
+#### Default Collectors
 
 The default collection (`array`) of `Collectors` provided by the SDK. New `Collectors` can be merged with the default `Collectors` to compose unique sequences of collection.
 
+#### Web Based Collectors
+
+TBC
```
Comment (MalcolmSmithUK) [inline README.md]: @robertdumitrescu INFO - Please complete this!

File: examples/Vanilla/WebBasedCollectorsCollection/webBasedCollectorsCollection.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+((tenancyId, rciSdk) => {
+  // Step 1: Configure your Transport with the tenancyId provided
+  const targetUrl = `https://target.domain/v1/${tenancyId}/stream`;
+  const transport = new rciSdk.Transport(targetUrl);
+
+  // Step 2.a: Capture your default collectors
+  const defaults = rciSdk.collector.defaultCollectors;
+
+  // Step 2.b: Capture WebFocusedCollectors
+  const {webFocusedCollectors} = rciSdk.collector;
```
Comment (MalcolmSmithUK) [inline examples/Vanilla/WebBasedCollectorsCollection/webBasedCollectorsCollection.js:10]: @robertdumitrescu READABILITY - This seems less `web focussed` and more `all the collectors`...  The name of the file is WebBasedCollectorsCollection.  The references are hard to follow for this example.

File: src/collector/WebBackEndCollector.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    return event;
+  }
+
+  /**
+     *
+     * @param {Object} currentPerformanceAPI
+     * @param {Object} oldPerformanceAPI
+     * @returns {null|Number}
+     */
+  static getResponseStart(currentPerformanceAPI, oldPerformanceAPI) {
```
Comment (MalcolmSmithUK) [inline src/collector/WebBackEndCollector.js]: @robertdumitrescu KISS - Mixing static methods in constructed instances should be avoided.  I'd suggest this could reference `this` (perhaps via binding) rather than feeding in as function arguments.  At it is here, it would appear the have the responsibility akin to a separate service.

File: src/collector/WebBackEndCollector.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    try {
+      if (currentPerformanceAPI) {
+        return Math.round(currentPerformanceAPI.responseStart);
```
Comment (MalcolmSmithUK) [inline src/collector/WebBackEndCollector.js]: @robertdumitrescu INFO - I'll assume this value is broken down into microseconds...?

File: src/collector/WebBackEndCollector.spec.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    eventType: 'state',
+    eventAction: 'load',
+    eventSource: 'products/shoes/1',
+    eventCategory: 'products/shoes',
+    deviceType: 'mobile'
+  };
+
+  const newPerfAPIResponse = [
+    {
+      responseStart: 555
```
Comment (MalcolmSmithUK) [inline src/collector/WebBackEndCollector.spec.js]: @robertdumitrescu COVERAGE - This is omitting the fact that the values are rounded.

File: src/collector/WebBackEndCollector.spec.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    beforeEach(() => {
+      webBackEndCollector = new WebBackEndCollector();
+    });
+
+    test('Return event with all mandatory fields', async () => {
+      /** @type {Event} */
+      const actualEvent = await webBackEndCollector.prepare(originalEvent);
+      const choppedActual = JSON.parse(JSON.stringify(actualEvent));
+      delete choppedActual.eventDuration1;
+      const choppedExpected = JSON.parse(JSON.stringify(expectedEvent));
```
Comment (MalcolmSmithUK) [inline src/collector/WebBackEndCollector.spec.js]: @robertdumitrescu KISS - You'll have to run this past me... this seems bonkers!

File: src/collector/WebPageLoadTimesCollector.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+   * @param {Object} oldPerformanceAPI
+   * @returns {null|Number}
+   */
+  static getDOMInteractive(currentPerformanceAPI, oldPerformanceAPI) {
+    try {
+      if (currentPerformanceAPI) {
+        return Math.round(currentPerformanceAPI.domInteractive);
+      } if (oldPerformanceAPI && oldPerformanceAPI.timing) {
+        /** We need to subtract due to the fact that is in EPOCH Time */
+        return oldPerformanceAPI.timing.domInteractive - oldPerformanceAPI.timing.navigationStart;
```
Comment (MalcolmSmithUK) [inline src/collector/WebPageLoadTimesCollector.js]: @robertdumitrescu BUG - I may be getting tired... but won't this result in a negative number?

File: src/collector/WebPageLoadTimesCollector.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+   * @param {Object} oldPerformanceAPI
+   * @returns {null|Number}
+   */
+  static getLoadEventStart(currentPerformanceAPI, oldPerformanceAPI) {
+    try {
+      if (currentPerformanceAPI) {
+        return Math.round(currentPerformanceAPI.loadEventStart);
+      } if (oldPerformanceAPI && oldPerformanceAPI.timing) {
+        /** We need to subtract due to the fact that is in EPOCH Time */
+        return oldPerformanceAPI.timing.loadEventStart - oldPerformanceAPI.timing.navigationStart;
```
Comment (MalcolmSmithUK) [inline src/collector/WebPageLoadTimesCollector.js]: @robertdumitrescu BUG - I may be getting tired... but won't this result in a negative number?

File: src/collector/WebPageLoadTimesCollector.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+   * @param {Object} oldPerformanceAPI
+   * @returns {null|Number}
+   */
+  static getDOMComplete(currentPerformanceAPI, oldPerformanceAPI) {
+    try {
+      if (currentPerformanceAPI) {
+        return Math.round(currentPerformanceAPI.domComplete);
+      } if (oldPerformanceAPI && oldPerformanceAPI.timing) {
+        /** We need to subtract due to the fact that is in EPOCH Time */
+        return oldPerformanceAPI.timing.domComplete - oldPerformanceAPI.timing.navigationStart;
```
Comment (MalcolmSmithUK) [inline src/collector/WebPageLoadTimesCollector.js]: @robertdumitrescu BUG - I may be getting tired... but won't this result in a negative number?

File: src/collector/WebPageLoadTimesCollector.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+   * @param {Object} oldPerformanceAPI
+   * @returns {null|Number}
+   */
+  static getLoadEventEnd(currentPerformanceAPI, oldPerformanceAPI) {
+    try {
+      if (currentPerformanceAPI) {
+        return Math.round(currentPerformanceAPI.loadEventEnd);
+      } if (oldPerformanceAPI && oldPerformanceAPI.timing) {
+        /** We need to subtract due to the fact that is in EPOCH Time */
+        return oldPerformanceAPI.timing.loadEventEnd - oldPerformanceAPI.timing.navigationStart;
```
Comment (MalcolmSmithUK) [inline src/collector/WebPageLoadTimesCollector.js]: @robertdumitrescu BUG - I may be getting tired... but won't this result in a negative number?

**Review by MalcolmSmithUK — CHANGES_REQUESTED** [review]
Part 1... To be continued!

File: examples/Vanilla/WebBasedCollectorsCollection/webBasedCollectorsCollection.js (hunk 3 above)
Comment (robertdumitrescu) [reply examples/Vanilla/WebBasedCollectorsCollection/webBasedCollectorsCollection.js:10]: The {} is doing object destructuring on rciSDK.collector...so those are only the 4 collectors that are web focused

File: src/collector/WebPageLoadTimesCollector.js (hunk 9 above)
Comment (robertdumitrescu) [reply src/collector/WebPageLoadTimesCollector.js]: the perf timing looks something like this:

`"{"connectStart":1602004610712,"navigationStart":1602004610706,"loadEventEnd":1602004614269,"domLoading":1602004612963,"secureConnectionStart":0,"fetchStart":1602004610712,"domContentLoadedEventStart":1602004613639,"responseStart":1602004612946,"responseEnd":1602004613410,"domInteractive":1602004613638,"domainLookupEnd":1602004610712,"redirectStart":0,"requestStart":1602004610751,"unloadEventEnd":0,"unloadEventStart":0,"domComplete":1602004614260,"domainLookupStart":1602004610712,"loadEventStart":1602004614261,"domContentLoadedEventEnd":1602004613640,"redirectEnd":0,"connectEnd":1602004610712}"`

so nav timing is basically the first metric that is happening. anything else happens after it.

File: src/collector/WebBackEndCollector.spec.js (hunk 7 above)
Comment (robertdumitrescu) [reply src/collector/WebBackEndCollector.spec.js]: Agreed, will look at another solution. that was a neat hack to see that it doesn't mess with the other values cause I had problems in keeping the initial object pristine, but yeah, is bonkers.

File: src/collector/WebBackEndCollector.spec.js (hunk 6 above)
Comment (robertdumitrescu) [reply src/collector/WebBackEndCollector.spec.js]: Good point.

File: src/collector/WebBackEndCollector.js (hunk 5 above)
Comment (robertdumitrescu) [reply src/collector/WebBackEndCollector.js]: Is in miliseconds. So on the new API, the value looks like this: `responseStart: 167.88500000257045`

File: src/collector/WebBackEndCollector.js (hunk 4 above)
Comment (robertdumitrescu) [reply src/collector/WebBackEndCollector.js]: Okay, this was a preference thing. I will change it. Doesn't bother me. I preferred it for the more clear API input/output but the collectors themselves are simple enough so is fine.

File: src/collector/WebBackEndCollector.spec.js (hunk 6 above)
Comment (robertdumitrescu) [reply src/collector/WebBackEndCollector.spec.js]: Done

File: src/collector/WebBackEndCollector.js (hunk 4 above)
Comment (robertdumitrescu) [reply src/collector/WebBackEndCollector.js]: Done

File: README.md (hunk 2 above)
Comment (robertdumitrescu) [reply README.md]: Done

File: src/collector/WebBackEndCollector.spec.js (hunk 7 above)
Comment (robertdumitrescu) [reply src/collector/WebBackEndCollector.spec.js]: Done

File: examples/Vanilla/WebBasedCollectorsCollection/webBasedCollectorsCollection.js (hunk 3 above)
Comment (MalcolmSmithUK) [reply examples/Vanilla/WebBasedCollectorsCollection/webBasedCollectorsCollection.js:10]: OK - your README does help make this clearer.

File: src/core/Trigger.helper.js [hunk 12]
```diff
@@ -13,6 +13,27 @@ export default class TriggerHelper {
     }
   }
 
+  static async defaultCondition () {
```
Comment (MalcolmSmithUK) [inline src/core/Trigger.helper.js:16]: @robertdumitrescu INFO - Not sure how this fits in?  How will it be used for existing implementations?

**Review by MalcolmSmithUK — CHANGES_REQUESTED** [review]

File: src/core/Trigger.helper.js (hunk 12 above)
Comment (robertdumitrescu) [reply src/core/Trigger.helper.js:16]: Instead of copying the exact same condition everywhere, we can use this one as a generic one, if there are some complex use cases we can still overwrite it.

**Review by MalcolmSmithUK — APPROVED** [review]
